### PR TITLE
WEA-15 Get new forecast of a location based on search query

### PR DESCRIPTION
**Pull request by ayush-shta**
Todo: Provide search suggestions.
@subash-poudel

### CODE REVIEW

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java [hunk 1]
```diff
@@ -57,8 +61,10 @@
 
   HomePresenterImpl(BaseContract.View view) {
     this.view = (HomeContract.View) view;
-    preferenceRepository = PreferenceRepositoryImpl.get();
 
+    geocodingAPIInterface = GeocodingAPIClient.getClient().create(GeocodingAPIInterface.class);
+
+    preferenceRepository = PreferenceRepositoryImpl.get();
     preferenceRepository.onPreferenceChangeListener(
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: refactor the change listener as there is no way to know the old temperature

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java [hunk 2]
```diff
@@ -57,8 +61,10 @@
 
   HomePresenterImpl(BaseContract.View view) {
     this.view = (HomeContract.View) view;
-    preferenceRepository = PreferenceRepositoryImpl.get();
 
+    geocodingAPIInterface = GeocodingAPIClient.getClient().create(GeocodingAPIInterface.class);
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: This should be done in a repository

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java [hunk 3]
```diff
@@ -130,7 +136,7 @@ private void startLocationUpdates() {
     if (ActivityCompat.checkSelfPermission(getContext(),
         Manifest.permission.ACCESS_FINE_LOCATION)
         != PackageManager.PERMISSION_GRANTED) {
-      throw new RuntimeException("Location permission not provided");
+      throw new RuntimeException("LocationCoordinates permission not provided");
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: refactor rename bug

Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java [hunk 4]
```diff
@@ -276,4 +286,13 @@ private void fetchHomeDetails() {
             new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)));
     dialog.setCancelable(false).show();
   }
+
+  @Override public boolean onQueryTextSubmit(String query) {
+    presenter.fetchNewLocation(query);
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java]: from the view's perspective, it is performing a search. Only the presenter should know that the search should be for a location. A appropriate name could be presenter.searchLocation(String query).

File: app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java [hunk 5]
```diff
@@ -276,4 +286,13 @@ private void fetchHomeDetails() {
...
+    return false;
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java]: Should this be true? Mention the reason for yes or no in comment.

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java (hunk 1 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java (hunk 2 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomePresenterImpl.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java (hunk 4 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java (hunk 5 above)
Comment (ayush-shta) [reply app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java]: :ok:

File: app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java [hunk 6]
```diff
@@ -276,4 +285,15 @@ private void fetchHomeDetails() {
             new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)));
     dialog.setCancelable(false).show();
   }
+
+  @Override public boolean onQueryTextSubmit(String query) {
+    presenter.searchLocation(query);
+    searchMenuItem.collapseActionView();
+    //return true to indicate that it has handled the submit request
```
Comment (subash-poudel) [inline app/src/main/java/com/ayush/weatherapp/home/HomeActivity.java:292]: comment not required. Is collapsing the action view desired. The search term might be lost.

**Pull request merged**